clisops/core/regrid.py

```diff
@@ -79,107 +235,241 @@ class Weights:
+            )
+        # Locstream for unstructured grids
+        locstream_in = False
+        locstream_out = False
+        if self.grid_in.type == "irregular":
+            locstream_in = True
+        if self.grid_out.type == "irregular":
+            locstream_out = True
 
         # Call xesmf.Regridder
+        if os.path.isfile(Path(weights_dir, self.filename).as_posix()):
+            reuse_weights = True
+        else:
+            reuse_weights = False
         return xe.Regridder(
             self.grid_in.ds,
             self.grid_out.ds,
             self.method,
-            periodic=periodic,
-            ignore_degenerate=ignore_degenerate,
+            periodic=self.periodic,
+            locstream_in=locstream_in,
+            locstream_out=locstream_out,
+            ignore_degenerate=self.ignore_degenerate,
             unmapped_to_nan=True,
+            filename=Path(weights_dir, self.filename).as_posix(),
+            reuse_weights=reuse_weights,
         )
+        # The default filename is important for later use, so reset it.
+        # xESMF writes weights to disk when filename is given and reuse_weights=False
+        # (latter is default) else it will create a default filename and weights can be manually
+        # written to disk with Regridder.to_netcdf(filename)
 
     def _reformat(self, format_from, format_to):
         raise NotImplementedError
 
     def _detect_format(self, ds):
         raise NotImplementedError
 
-    def generate_id(self):
+    def _generate_id(self):
         # A unique id could maybe consist of:
         #  - method
         #  - hash/checksum of input and output grid (lat, lon, lat_bnds, lon_bnds)
-        return
+        peri_dict = {True: "peri", False: ""}
+        ignore_degenerate_dict = {None: "", True: "skip_degen", False: ""}
+        wid = "_".join(
+            filter(
+                None,
+                [
+                    self.grid_in.hash,
+                    self.grid_out.hash,
+                    peri_dict[self.periodic],
+                    ignore_degenerate_dict[self.ignore_degenerate],
+                    self.method,
+                ],
+            )
+        )
+        return wid
 
+    @check_weights_dir
     def save_to_cache(self):
         # Create folder dependent on id
         # Store weight file in xESMF-format
         # Further information will be stored in a json file in the same folder
-        # weightfile=regridderHR.to_netcdf(regridder.filename)
-        raise NotImplementedError
-
-    def load_from_cache(self, id):
-        # Identify folder by id
-        # Load weightfile with xesmf.Regridder
+        # todo: parallel write mechanism to json file and weights file?
+        # todo: store the horizontal input and output grid in an extra netcdf file?
+        if self.id not in weights_dic:
+            try:
+                grid_in_source = self.grid_in.ds.encoding["source"]
```

Comment:
Could these be self.grid_in.ds.encoding.get("source") etc. instead, so that they are set to None if they are not found?